OpenRA.Mods.Common/Traits/Buildings/Building.cs

```diff
@@ -213,18 +222,35 @@ public void NotifyBuildingComplete(Actor self)
 
 		public void Selling(Actor self)
 		{
+			if (Info.RemoveSmudgesOnSell)
+				RemoveSmudges();
+
 			BuildComplete = false;
 		}
 
 		public void Sold(Actor self) { }
 
 		public void BeforeTransform(Actor self)
 		{
+			if (Info.RemoveSmudgesOnTransform)
+				RemoveSmudges();
+
 			foreach (var s in Info.UndeploySounds)
 				Game.Sound.PlayToPlayer(self.Owner, s, self.CenterPosition);
 		}
 
 		public void OnTransform(Actor self) { }
 		public void AfterTransform(Actor self) { }
+
+		public void RemoveSmudges()
+		{
+			var smudgeLayers = self.World.WorldActor.TraitsImplementing<SmudgeLayer>();
+
+			foreach (var occupiedCell in OccupiedCells())
+			{
+				foreach (var smudgeLayer in smudgeLayers)
+					smudgeLayer.RemoveSmudge(occupiedCell.First);
+			}
```

Comment:
You may wish to switch the two loops here so you only have to enumerate the smudge layers once.

```c#
foreach (var smudgeLayer in smudgeLayers)
    foreach (var occupiedCell in OccupiedCells())
        ...
```